fix: NetworkTransform initial synchronization, parenting, smooth transitions between transform spaces, and UI updates

NoelStephensUnity opened the pull request:
## Parenting & Smooth Transform Space Transitions
This PR contains a handful of `NetworkTransform` fixes pertaining to parenting, the initial synchronizing parented `NetworkObject`s with newly/late joined clients, issues with smooth transform space transitions (_specific to parenting_) while interpolation is enabled, and provides the ability to tick synchronize nested and parented `NetworkTransform` instances.

**Side Note:** #3039 is the example project to that demonstrates these updates and includes additional (_and useful_) helper classes.

The below described properties allow for seamless transform space parenting and transitions when interpolation is enabled (_primarily focusing on the owner authoritative motion model_), and only requires setting the parent of the owned `NetworkObject` (i.e. no more predicting RTT between the server and client owner) to accomplish this. 
**Side Note:** This is only for non-Rigidbody spawned instances. There will be a follow up PR to handle the Rigidbody approach that will still leverage from some of the below properties, but there is a different dynamic involved with PhysX based "parenting" (i_n reality there is no such thing as local space in the PhysX simulation so using Joints combined with tick synchronized children is the better approach_).

### `NetworkTransform.SwitchTransformSpaceWhenParented`
![image](https://github.com/user-attachments/assets/489c65ea-d00d-45d8-81ce-c305e4ffa953)
When enabled, the "Tick Sync Children" property is automatically enabled to assure any already parented children will be synchronized properly as well as it handles tick synchronizing nested `NetworkTransform` components. This also automatically updates owner authoritative's `NetworkTransform.InLocalSpace` setting, handles updating/converting each `LinearBufferInterpolator` entries to the proper transform space relative to the current, previous, or current and previous (if doing a `NetworkObject` to `NetworkObject` transition) parent(s).

This does not apply to `NetworkTransform` component(s) paired with `NetworkRigidbody` component(s), as that will be the next PR update for v2.0.0.

### `NetworkTransform.TickSyncChildren`
When enabled, nested `NetworkTransform` components will be synchronized with the root `NetworkTransform` or in the event they are all nested under a root `GameObject` with `NetworkObject` component the first `NetworkTransform` in that hierarchy. 
#### How this works:
When the root `NetworkTransform` (_on the authoritative instance_) detects a state update needs to be sent, it will then have all nested and/or child (parented) `NetworkTransforms` (_as long as the client has authority over them_) send state updates. This helps to prevent a potential issue where nested or child `NetworkTransform` components could get into a state where they are all sending updates on different ticks which can lead to visual "jittering" between the associated `GameObject` transforms.
_This property is automatically enabled when `NetworkTransform.SwitchTransformSpaceWhenParented` is enabled._

### `NetworkObject.AllowOwnerToParent` 
![image](https://github.com/user-attachments/assets/8d945eee-f843-4cc1-bfe2-519a31191b0f)
This is primarily a property for client-server network topology based projects that might use an owner authoritative motion model (_typically for player avatars/prefabs_). Enabling this property allows the client owner to immediately parent the owned object under another `NetworkObject` in order to assure there is no latency between the parenting event and `NetworkTransform` updates. This is especially important when handling `NetworkObject` to `NetworkObject` parenting when using a client-server network topology in order to maintain smooth transitions on non-authoritative instances.
_Note: Setting this property when using a distributed authority network topology has no effect since client owners already have the authority to handle immediate local parenting._

### `NetworkObject.SyncOwnerTransformWhenParented`
![image](https://github.com/user-attachments/assets/2ad01a32-0fca-48c9-b68c-44480189ed5b)
When disabled, this property will not apply the server-side transform values received via the `ParentSyncMessage` on the client owner side. This is more specific to owner authoritative motion models (_i.e. `NetworkTransform` with owner authority_) but can be useful under other scenarios where having the client owner not applying the server-side's transform state is required. Disabling this property impacts non-authoritative instances in the following ways (depending upon the network topology being used):
- Client-server: The non-authoritative instances will always apply the server's transform state. _The server's transform state leads the non-authoritative instance(s) state(s) by roughly 1 network tick relative to the client owner's transform state._
- Distributed authority: The non-authoritative instances will always apply the owner's transform state. Since the client has authority. Because of this, enabling this property has no significant impact.
When enabled (default),  client owner will apply the server's transform state when receiving the `ParentSyncMessage` from the server (but has no impact when using a distributed authority network topology).

## UI Updates & `NetcodeEditorBase`
This PR also includes a new `NetcodeEditorBase` editor that can be used with any `MonoBehaviour` derived component and helps to organize inspector view properties by child class generation.  The `NetworkTransform`, `NetworkRigidbodyBase`, and `NetworkManager` editors now derive from `NetcodeEditorBase` so any child derived from one of these three components will provide this separation of child generation relative properties. When used in their original form (i.e. not a component derived from) it will show only the properties and not apply the foldout group.

### Folder Grouping of Each Generational Child
The `NetcodeEditorBase`, used by `NetworkTransform`, `NetworkRigidbodyBase`, and `NetworkManager` custom editors, provides a clean separation between the base class and any derived child.

### `NetworkTransform` Example:
The `NetworkTransform` derived both child generation properties are collapsed:
![image](https://github.com/user-attachments/assets/c5e38c08-81ab-482e-849b-501a7527b9cd)
Expanding the last generation child, `MoverScriptNoRigidbody`, properties:
![image](https://github.com/user-attachments/assets/190b3b83-ffb4-48f6-b869-cd3f680c5d4e)
Expanding the `NetworkTransform` parent class properties:
![image](https://github.com/user-attachments/assets/6a217544-51e5-499f-9bfd-3a04bd6ee530)

### `NetworkManager` Example:
For those who like to store additional netcode and/or project specific settings in the same place, deriving from the `NetworkManager` while using `NetcodeEditorBase` for the `ExtendedNetworkManagerEditor`:
Collapsed:
![image](https://github.com/user-attachments/assets/e620ac8d-52c2-4e39-82bb-b11d69168483)
Fully expanded:
![image](https://github.com/user-attachments/assets/6d9e8163-b84c-4655-8718-982810142df5)

### Multi-Object Compatible
The `NetcodeEditorBase` is Multi-Object compatible. Below is a prefab with several children selected that have the same `PlayerBallMotion : NetworkTransform` component:
![image](https://github.com/user-attachments/assets/fd15b6a5-9436-43c6-b223-72835a20ae92)
When the `NetworkTransform` properties are expanded, you get the same custom editor view for all selected objects:
![image](https://github.com/user-attachments/assets/059eee8f-0440-44dc-b757-55aea90cecde)



[MTTB-66](https://jira.unity3d.com/browse/MTTB-66)
[MTTB-6](https://jira.unity3d.com/browse/MTTB-6)

fix: #2959
fix: #2913
fix: #2856
fix: #2843
fix: #2842


## Changelog
- Added `NetworkTransform.SwitchTransformSpaceWhenParented` property that, when enabled, will handle the world to local, local to world, and local to local transform space transitions when interpolation is enabled.
- Added `NetworkTransform.TickSyncChildren` that, when enabled, will tick synchronize nested and/or child `NetworkTransform` components to eliminate any potential visual jittering that could occur if the `NetworkTransform` instances get into a state where their state updates are landing on different network ticks.
- Added `NetworkObject.AllowOwnerToParent` property to provide the ability to allow clients to parent owned objects when running in a client-server network topology.
- Added `NetworkObject.SyncOwnerTransformWhenParented` property to provide a way to disable applying the server's transform information in the parenting message on the client owner instance which can be useful for owner authoritative motion models.
- Added `NetcodeEditorBase` editor helper class to provide easier modification and extension of the SDK's components.
- Fixed issue with newly/late joined clients and `NetworkTransform` synchronization of parented `NetworkObject` instances.
- Fixed issue with smooth transitions between transform spaces when interpolation is enabled (requires `NetworkTransform.SwitchTransformSpaceWhenParented` to be enabled).
- Changed `NetworkTransformEditor` so it now derives from `NetcodeEditorBase`.
- Changed `NetworkRigidbodyBaseEditor` so it now derives from `NetcodeEditorBase`.
- Changed `NetworkManagerEditor` so it now derives from `NetcodeEditorBase`.

## Testing and Documentation
- Includes integration tests.
- Documentation additions are necessary.


<!--  Uncomment and mark items off with a * if this PR deprecates any API:
### Deprecated API
- [ ] An `[Obsolete]` attribute was added along with a `(RemovedAfter yyyy-mm-dd)` entry.
- [ ] An [api updater] was added.
- [ ] Deprecation of the API is explained in the CHANGELOG.
- [ ] The users can understand why this API was removed and what they should use instead.
-->

NoelStephensUnity committed:
fix

This fixes a synchronization issue with owner authoritative instances when running in client-server mode and parenting. When the server sends a parent sync message back with the position of the client's owned object it is out of sync with the client's position by the time the message is received. This fix gives users the option to not synchronize the transform position of a NetworkObject when using an owner authoritative motion mode.

NoelStephensUnity committed:
fix

This fix provides users with the option to have child `NetworkTransform`(s) tick synchronized with their parent. When the parent has `TickSyncChildren` enabled, any child `NetworkTransform`(s) that the local client has authority over will force a synchronization of their transform state. When a client is not the authority of the parent but it is the authority of one or more children, then when the parent sends a state update it will force the child/children the client has authority over to send a state update. Children will still send state updates if the transform has a delta that exceeds an axis threshold and the parent has yet to send a synchronization, but if a parent sends a synchronization within the same tick period the child/children will still send another state update.

NoelStephensUnity committed:
wip

Better approach, but still flawed.
I think converting any remaining items in the interpolators when the transform space changes will yield consistent results.

NoelStephensUnity committed:
fix

Seamless transform space transitions when parenting.

NoelStephensUnity committed:
update

We no longer need to send lossy scale since we defer applying the initial synchronization data until the non-authority side has finished synchronizing.

NoelStephensUnity committed:
fix

Fixing some minor issues with nested `NetworkTransforms` (i.e. under the same `NetworkObject`) and adding additional comments.
Reverting back to just checking `InLocalSpace` (as opposed to the interpolator's) when applying a position on the non-authority side.

NoelStephensUnity committed:
test fix

Fixing some issues with `NestedNetworkTransformTests` assets to account for some past updates.
Fixing an issue with `NetworkTransformTests` where it needs to not just advance the tick (for updates) but needs to also advance to the next tick.

NoelStephensUnity committed:
fix

caught scenario where `NetworkObject.NetworkTransforms` could not be set properly.

NoelStephensUnity committed:
fix

This fixes a minor "blip" when transitioning from world to local space,

NoelStephensUnity committed:
test

disabling two tests that need to be reviewed by Kitty and/or re-written.

NoelStephensUnity committed:
fix

Updating all NetworkTransform instances that belong to a specific NetworkObject as opposed to just the root.
Also not updating when not spawned or disabled NetworkTransforms.

NoelStephensUnity committed:
test

Making test asset adjustments for recent updates in NetworkTransform.

NoelStephensUnity committed:
update

Adding change log entry.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/nested-networktransforms-not-updating

NoelStephensUnity committed:
update

mid-point check-in

NoelStephensUnity committed:
test

Updating interpolation test to recent NetworkTransform updates.

NoelStephensUnity committed:
Merge branch 'fix/nested-networktransforms-not-updating' into fix/owner-object-parenting-no-sync-transform-option

NoelStephensUnity committed:
update

wip towards getting a cleaner divide between the two modes.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/owner-object-parenting-no-sync-transform-option

NoelStephensUnity committed:
update

Some adjustments to when nested children check their own state updates or exit early.

NoelStephensUnity committed:
update

Cleaning up some Rigidbody references and making Rigidbody & Rigidbody2D accessible via NetworkRigidbody and NetworkRigidbody2D.

NoelStephensUnity committed:
update

Adding an internal InternalOnNetworkObjectParentChanged method that gets invoked ahead of OnNetworkObjectParentChanged so user code can't override any NGO components that need this kind of notification.

NoelStephensUnity committed:
update

Adding InternalOnNetworkSessionSynchronized to be invoked by NGO components to assure they don't get overridden by user script.

NoelStephensUnity committed:
update

Renaming some internal properties that are common names and would conflict with serialization if a user defined them in a derived class.

Making sure both the position and rotation interpolators are configured for the transform space when reset.

Adding the ability to change the parent locally when running with a client-server network topology.
Adding SwitchTransformSpaceWhenParented property to the NetworkTransformEditor.
Shifting all of the non-authority's final synchronization logic to occur either once a client's initial connection synchronization is complete or when all of the NetworkObject's components have run through the spawn process.

NoelStephensUnity committed:
update

Adding some improved and needed UI updates to NetworkBehaviour derived components and NetworkManager.
This includes a new NetcodeEditorBase class that allows users to more easily create customized, multi-generation, components that have their properties organized based on the generation of each child.

NoelStephensUnity committed:
fix

Wrapping the NetworkRigidbodyBaseEditor within physics defines.

NoelStephensUnity committed:
style

>.<

(sigh) null checks can be simplified style/standards update.

NoelStephensUnity committed:
update

switching from TT as a class to TT as a MonoBehaviour.

NoelStephensUnity committed:
fix

This fixes an issue where a server instance parenting with AllowOwnerToParent set would send to itself.
Adding check within the NetworkObject parenting logic to pass through to the server portion of sending the parenting message.

NoelStephensUnity committed:
test

adding some validation tests for the updates made in #3013

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/owner-object-parenting-no-sync-transform-option

NoelStephensUnity committed:
update

adding changelog entries

NoelStephensUnity committed:
fix

This fixes the two failing tests within NetworkTransformAnticipationTest and updates AnticipatedNetworkTransform to account for #3013's updates/changes.

NoelStephensUnity committed:
fix

Removing the added base.OnInspectorGUI() call (was causing duplicated properties)

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/owner-object-parenting-no-sync-transform-option

danni-lu commented on com.unity.netcode.gameobjects/Editor/NetworkManagerEditor.cs:
I have a little comment here. I like the foldout idea, but I think maybe we can take the call-to-action buttons out so users can access them without expanding the foldout content, especially during runtime when there's nothing much within the foldout section. I made a quick mock here! Please lemme know if this makes sense! 

Side note: would it make sense to show network manager properties during runtime? will people need this information?

![image](https://github.com/user-attachments/assets/43bed7c8-af97-46df-a6dd-1cdca5b92c38)

NoelStephensUnity replied:
Awesome feedback... and good points.
Made the suggested changes and pushed the update.
👍

NoelStephensUnity committed:
update

Assure the callouts and call-to-action buttons show up outside of the Foldout Group when extending NetworkManager.
Assure that the NetworkManager has no Foldout Group when it is not extended.
